File: backend/src/lib/pagination.ts
const DEFAULT_LIMIT = 20;
const MAX_LIMIT = 100;

export interface Pagination {
  page: number;
  limit: number;
  offset: number;
}

function parseIntOr(value: unknown, fallback: number): number {
  const parsed = Number.parseInt(String(value ?? ''), 10);
  return Number.isFinite(parsed) ? parsed : fallback;
}

export function parsePagination(query: Record<string, unknown>): Pagination {
  const page = Math.max(1, parseIntOr(query.page, 1));
  const limit = Math.min(MAX_LIMIT, Math.max(1, parseIntOr(query.limit, DEFAULT_LIMIT)));
  return { page, limit, offset: (page - 1) * limit };
}

export function paginated<T>(data: T[], pagination: Pagination, total: number) {
  return {
    data,
    pagination: {
      page: pagination.page,
      limit: pagination.limit,
      total,
      total_pages: Math.max(1, Math.ceil(total / pagination.limit)),
    },
  };
}

File: app/lib/models/pagination.dart
class PageInfo {
  const PageInfo({required this.page, required this.limit, required this.total, required this.totalPages});

  final int page;
  final int limit;
  final int total;
  final int totalPages;

  factory PageInfo.fromJson(Map<String, dynamic> json) => PageInfo(
        page: json['page'] as int,
        limit: json['limit'] as int,
        total: json['total'] as int,
        totalPages: json['total_pages'] as int,
      );
}

class Paginated<T> {
  const Paginated({required this.data, required this.pageInfo});

  final List<T> data;
  final PageInfo pageInfo;

  factory Paginated.fromJson(Map<String, dynamic> json, T Function(Map<String, dynamic>) fromJson) => Paginated(
        data: (json['data'] as List).cast<Map<String, dynamic>>().map(fromJson).toList(),
        pageInfo: PageInfo.fromJson(json['pagination'] as Map<String, dynamic>),
      );
}
